File: libs/power-log-parser/src/lib/parsers/battlegrounds-buddy-gained-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider } from '../game-event';
import { Node, NodeType, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class BattlegroundsBuddyGainedParser implements ActionParser {
	readonly ParserName = 'BattlegroundsBuddyGainedParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			this.StateFacade.IsBattlegrounds() &&
			node.Type === NodeType.TagChange &&
			(node.Object as TagChange).Name === (GameTag.BACON_PLAYER_NUM_HERO_BUDDIES_GAINED as number)
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const hero = this.GameState.CurrentEntities.get(tagChange.Entity);
		if (hero == null) {
			return null;
		}

		const heroCardId = hero.CardId;
		const heroes = [...this.GameState.CurrentEntities.values()]
			.filter((entity) => entity.CardId === heroCardId)
			.filter((entity) => entity.GetTag(GameTag.BACON_PLAYER_NUM_HERO_BUDDIES_GAINED) >= tagChange.Value);
		if (heroes.length > 0) {
			return null;
		}

		if (hero?.CardId != null && hero.CardId.length > 0 && !hero.IsBaconBartender() && tagChange.Value >= 1) {
			return [
				GameEventProvider.Create(
					tagChange.TimeStamp,
					'BATTLEGROUNDS_BUDDY_GAINED',
					() => ({
						Type: 'BATTLEGROUNDS_BUDDY_GAINED',
						Value: {
							CardId: hero.CardId,
							PlayerId: hero.GetTag(GameTag.PLAYER_ID, 0),
							TotalBuddies: tagChange.Value,
						},
					}),
					false,
					node,
				),
			];
		}
		return null;
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BattlegroundsBuddyGainedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Linq;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using static HearthstoneReplays.Events.CardIds;

namespace HearthstoneReplays.Events.Parsers
{
    public class BattlegroundsBuddyGainedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public BattlegroundsBuddyGainedParser(ParserState ParserState, StateFacade stateFacade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = stateFacade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && StateFacade.IsBattlegrounds()
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.BACON_PLAYER_NUM_HERO_BUDDIES_GAINED;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var hero = GameState.CurrentEntities[tagChange.Entity];
            if (hero == null)
            {
                return null;
            }

            // Add a safeguard to avoid duplicated info
            var heroCardId = hero.CardId;
            var heroes = GameState.CurrentEntities.Values
                .Where(entity => entity.CardId == heroCardId)
                .Where(entity => entity.GetTag(GameTag.BACON_PLAYER_NUM_HERO_BUDDIES_GAINED) >= tagChange.Value)
                .ToList();
            if (heroes.Count > 0)
            {
                return null;
            }

            // The value is set to 0 when rotating the entities it seems
            if (hero?.CardId != null && !hero.IsBaconBartender() && tagChange.Value >= 1)
            {
                return new List<GameEventProvider> {  GameEventProvider.Create(
                    tagChange.TimeStamp,
                    "BATTLEGROUNDS_BUDDY_GAINED",      
                    () => new GameEvent
                    {
                        Type = "BATTLEGROUNDS_BUDDY_GAINED",
                        Value = new
                        {
                            CardId = hero.CardId,
                            PlayerId = hero.GetTag(GameTag.PLAYER_ID, 0),
                            TotalBuddies = tagChange.Value,
                        }
                    },
               false,
               node) };
            }
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
